src/main/java/org/students/simplebitcoinnode/service/MiningService.java

```diff
@@ -0,0 +1,97 @@
+import org.students.simplebitcoinnode.dto.PoolInitiationBlockMetadataDTO;
+import org.students.simplebitcoinnode.dto.PoolMiningProposalDTO;
+import org.students.simplebitcoinnode.entity.AdjacentNode;
+import org.students.simplebitcoinnode.entity.Block;
+import org.students.simplebitcoinnode.entity.MinerPublicKey;
+import org.students.simplebitcoinnode.event.MineBlockEvent;
+import org.students.simplebitcoinnode.repository.AdjacentNodeRepository;
+import org.students.simplebitcoinnode.repository.BlockRepository;
+import org.students.simplebitcoinnode.util.DTOMapperWrapper;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.logging.Logger;
+
+@Service
+public class MiningService {
+    Logger logger = Logger.getLogger(NodeRegistrationService.class.getName());
+    private final DTOMapperWrapper dtoMapperWrapper;
+    private final AdjacentNodeRepository adjacentNodeRepository;
+    private final ApplicationEventPublisher applicationEventPublisher;
+    private final BlockchainMiningPoolConfig blockchainMiningPoolConfig;
+    private final BlockRepository blockRepository;
+
+    public MiningService(DTOMapperWrapper dtoMapperWrapper, AdjacentNodeRepository adjacentNodeRepository, ApplicationEventPublisher applicationEventPublisher, BlockchainMiningPoolConfig blockchainMiningPoolConfig, BlockRepository blockRepository) {
+        this.dtoMapperWrapper = dtoMapperWrapper;
+        this.adjacentNodeRepository = adjacentNodeRepository;
+        this.applicationEventPublisher = applicationEventPublisher;
+        this.blockchainMiningPoolConfig = blockchainMiningPoolConfig;
+        this.blockRepository = blockRepository;
+    }
+    
+    /**
+     * Proposes a new mining operation.
+     *
+     * @param proposalDTO The data transfer object containing the details of the proposal.
+     * @throws IllegalArgumentException If the expected pool size is larger than the allowed size or if the previous block hash does not represent a block with the longest chain.
+     */
+    public void propose(PoolMiningProposalDTO proposalDTO){
+        // Check that amount of expected pool size is lower than allowed one.
+        Integer maximumPoolRequestsSize = blockchainMiningPoolConfig.getMaximumPoolRequests();
+        Integer expectedPoolSize = proposalDTO.getExpectedPoolSize();
+        if (maximumPoolRequestsSize < expectedPoolSize) throw new IllegalArgumentException("Expected pool size is bigger than allowed size. Allowed size is: " + maximumPoolRequestsSize);
+
+        // Check that the previous block hash represents a block with the longest chain.
+        String blockWithLongestChainHash= blockRepository.findBlockWithLongestChain().getHash();
+        String previousBlockHash = proposalDTO.getBlock().getPreviousHash();
+        if (!previousBlockHash.equals(blockWithLongestChainHash)) throw new IllegalArgumentException("Passed previous block has is invalid!");
```

Comment:
Don't do validation in service layer but instead create separate ConstraintValidator classes with corresponding annotations. See `entity.validation` package for examples.